providers: add Cloudflare AI Gateway

elithrar opened the pull request:
This PR:

- [x] Adds Cloudflare's AI Gateway (https://developers.cloudflare.com/ai-gateway/) as a provider to opencode
- [x] Updates the Providers (https://opencode.ai/docs/providers/) docs
- [x] Links to the corresponding models.dev `cloudflare-ai-gateway` provider to autoload models - PR is here: https://github.com/sst/models.dev/pull/476 
- [x] Validate model selection works

elithrar commented:
Let me see what’s up with the format failure when I’m back tomorrow.

rekram1-node commented:
@elithrar You can ignore the ci failure, it'll autoformat anything not formatted after being merged anyways

rekram1-node commented:
This looks good to me, models.dev has been updated and given it seems this works for you, it's safe to merge but I'll hold off till you are ready

elithrar commented:
Thanks! I’m seeing a few issues with some Workers AI models - want to debug
that before this goes live. Suspect the model naming format isn’t working.

On Sat, Dec 6, 2025 at 15:24 Aiden Cline ***@***.***> wrote:

> *rekram1-node* left a comment (sst/opencode#5174)
> <https://github.com/sst/opencode/pull/5174#issuecomment-3621082933>
>
> This looks good to me, models.dev has been updated and given it seems
> this works for you, it's safe to merge but I'll hold off till you are ready
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sst/opencode/pull/5174#issuecomment-3621082933>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEQ4BYNAA4C72PN2BBVNL4AM3QHAVCNFSM6AAAAACOH6K3SGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMRRGA4DEOJTGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

elithrar commented:
Fixed the merge conflict. Waiting on https://github.com/sst/models.dev/pull/514 to be merged in.

rekram1-node commented:
@elithrar merged

elithrar committed:
feat(cloudflare-ai-gateway): add Cloudflare AI Gateway provider with Unified Billing support

- Use cf-aig-authorization header for authenticated gateways
- Strip Authorization header via custom fetch to prevent auth errors
- Add getModel function using sdk.chat() for OpenAI-compatible endpoint
- Support API token from env var or /connect command
- Update docs with setup instructions

rekram1-node commented:
Btw (and we prolly need a cleaner way of doing stuff) but since opencode is still using AI Sdk v5 and the dynamic provider loading just grabs latest, I think if it tries loading `ai-gateway-provider` it will fail.

You can add the package into packages/opencode/package.json and pin the version to v5 compat and then update packages/opencode/..../provider/provider.ts to load it (youll see the logic im referring to at the top of the file)

elithrar commented:
Got it, great - let me do that.

testing this again but at some point probably need to support a
—models-dev-endpoint flag for overriding with a localhost instance of
models.dev too so it’s easier to validate e2e

On Tue, Dec 23, 2025 at 14:43 Aiden Cline ***@***.***> wrote:

> *rekram1-node* left a comment (sst/opencode#5174)
> <https://github.com/sst/opencode/pull/5174#issuecomment-3687788650>
>
> Btw (and we prolly need a cleaner way of doing stuff) but since opencode
> is still using AI Sdk v5 and the dynamic provider loading just grabs
> latest, I think if it tries loading ai-gateway-provider it will fail.
>
> You can add the package into packages/opencode/package.json and pin the
> version to v5 compat and then update
> packages/opencode/..../provider/provider.ts to load it (youll see the logic
> im referring to at the top of the file)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sst/opencode/pull/5174#issuecomment-3687788650>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEQ4GBXIE3VIKZGMEZGCT4DGLMLAVCNFSM6AAAAACOH6K3SGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMOBXG44DQNRVGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

rekram1-node commented:
Yeah that's a good idea

elithrar commented:
merge + deploy https://github.com/sst/models.dev/pull/560 first

elithrar committed:
Merge branch 'dev' into cloudflare-ai-gateway

rekram1-node commented:
models.dev pr has been merged + deployed

rekram1-node commented on packages/opencode/src/provider/provider.ts:
believe we need to add ur provider here

rekram1-node replied:
I could be wrong tho, are the models working w/ tool calls for you too? 

The error would happen during tool call if the v6 problem happens for CF

elithrar replied:
we're using `@ai-sdk/open-ai-compatible` (per our models.dev config) so this should just work

rekram1-node replied:
Oh true, sorry the 2 cloudlfare things in models.dev tripped me up

elithrar commented:
* re: v5 compat - this should be OK - `@ai-sdk/openai-compatible ` is pinned to `v1.0.27` (v5 compat) - ?

elithrar commented:
I still crash the TUI on any model selection for Cloudflare AI Gateway: 

The stack trace isn't useful as it's stuck in the spinning/TUI vs. the underlying provider exception:

```
Error: [Reconciler] Unknown component type: spinner
    at createElement (../../node_modules/.bun/@opentui+solid@0.0.0-20251211-4403a69a+897ebc1644d3edd0/node_modules/@opentui/solid/index.js:521:17)
    at children (src/cli/cmd/tui/component/prompt/index.tsx:874:24)
    at <anonymous> (../../node_modules/.bun/solid-js@1.9.10/node_modules/solid-js/dist/dev.js:1568:19)
    at runComputation (../../node_modules/.bun/solid-js@1.9.10/node_modules/solid-js/dist/dev.js:742:22)
    at updateComputation (../../node_modules/.bun/solid-js@1.9.10/node_modules/solid-js/dist/dev.js:724:3)
    at readSignal (../../node_modules/.bun/solid-js@1.9.10/node_modules/solid-js/dist/dev.js:660:67)
    at <anonymous> (../../node_modules/.bun/@opentui+solid@0.0.0-20251211-4403a69a+897ebc1644d3edd0/node_modules/@opentui/solid/index.js:148:62)
    at runComputation (../../node_modules/.bun/solid-js@1.9.10/node_modules/solid-js/dist/dev.js:742:22)
    at updateComputation (../../node_modules/.bun/solid-js@1.9.10/node_modules/solid-js/dist/dev.js:724:3)
    at createRenderEffect (../../node_modules/.bun/solid-js@1.9.10/node_modules/solid-js/dist/dev.js:240:75)...
```

elithrar commented:
OK, we have this working. I overrode the models based on the latest models.dev/api.json + fixed the way we call the `openai-compatible` SDK (`.languageModel`).

Local builds from this branch work as expected: 

```sh
# Set environment variables
export CLOUDFLARE_ACCOUNT_ID="your-account-id"
export CLOUDFLARE_GATEWAY_ID="your-gateway-name"
export CLOUDFLARE_API_TOKEN="your-api-token"

# OpenAI via Cloudflare AI Gateway
./dist/opencode-darwin-arm64/bin/opencode run --model "cloudflare-ai-gateway/openai/gpt-4o" "explain what opencode does"

# Anthropic via Cloudflare AI Gateway  
./dist/opencode-darwin-arm64/bin/opencode run --model "cloudflare-ai-gateway/anthropic/claude-sonnet-4-5" "write a hello world in rust"

# Workers AI
./dist/opencode-darwin-arm64/bin/opencode run --model "cloudflare-ai-gateway/workers-ai/@cf/qwen/qwen2.5-coder-32b-instruct" "fix the bug in main.ts"

# List available models
./dist/opencode-darwin-arm64/bin/opencode models | grep cloudflare-ai-gateway
```

```sh
➜ ./dist/opencode-darwin-arm64/bin/opencode run --model "cloudflare-ai-gateway/anthropic/claude-sonnet-4-5" "write a hello world in rust"

I'll create a simple "Hello, World!" program in Rust for you.

|  Write    packages/opencode/hello.rs

I've created `hello.rs` with a simple Hello World program. You can run it with:

rustc hello.rs && ./hello

Or if you have `cargo` and want to run it directly:

rustc hello.rs && ./hello
```

elithrar committed:
Merge branch 'dev' into cloudflare-ai-gateway

rekram1-node commented:
Thanks!

ahosker commented:
Is OpenCode Zen to be added to CloudFlare AI Gateway @elithrar /or/ with this added is their a better way to do this?
<img width="911" height="160" alt="image" src="https://github.com/user-attachments/assets/cd23e57f-38fc-462d-a4ec-bcffc0d10406" />

elithrar commented:
@ahosker right now you’d want to set it up as a custom provider. Can’t speak to whether we have plans otherwise.
